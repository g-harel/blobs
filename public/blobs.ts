import {genFromOptions} from "../internal/gen";
import {renderPath} from "../internal/render/svg";
import {renderPath2D} from "../internal/render/canvas";
import {mapPoints} from "../internal/util";
import {checkBlobOptions, checkCanvasOptions, checkSvgOptions} from "../internal/check";

export interface BlobOptions {
    // A given seed will always produce the same blob.
    // Use `Math.random()` for pseudorandom behavior.
    seed: string | number;
    // Actual number of points will be `3 + extraPoints`.
    extraPoints: number;
    // Increases the amount of variation in point position.
    randomness: number;
    // Size of the bounding box.
    size: number;
}

export interface CanvasOptions {
    // Coordinates of top-left corner of the blob.
    offsetX?: number;
    offsetY?: number;
}

export interface SvgOptions {
    fill?: string; // Default: "#ec576b".
    stroke?: string; // Default: "none".
    strokeWidth?: number; // Default: 0.
}

export const canvasPath = (blobOptions: BlobOptions, canvasOptions: CanvasOptions = {}): Path2D => {
    try {
        checkBlobOptions(blobOptions);
        checkCanvasOptions(canvasOptions);
    } catch (e) {
        throw `(blobs2): ${e}`;
    }
    return renderPath2D(
        mapPoints(genFromOptions(blobOptions), ({curr}) => {
            curr.x += canvasOptions.offsetX || 0;
            curr.y += canvasOptions.offsetY || 0;
            return curr;
        }),
    );
};

export const svg = (blobOptions: BlobOptions, svgOptions: SvgOptions = {}): string => {
    try {
        checkBlobOptions(blobOptions);
        checkSvgOptions(svgOptions);
    } catch (e) {
        throw `(blobs2): ${e}`;
    }
    const path = svgPath(blobOptions);
    const size = Math.floor(blobOptions.size);
    const fill = svgOptions.fill === undefined ? "#ec576b" : svgOptions.fill;
    const stroke = svgOptions.stroke === undefined ? "none" : svgOptions.stroke;
    const strokeWidth = svgOptions.strokeWidth === undefined ? 0 : svgOptions.strokeWidth;
    return `
<svg width="${size}" height="${size}" viewBox="0 0 ${size} ${size}" xmlns="http://www.w3.org/2000/svg">
    <path stroke="${stroke}" stroke-width="${strokeWidth}" fill="${fill}" d="${path}"/>
</svg>`.trim();
};

export const svgPath = (blobOptions: BlobOptions): string => {
    try {
        checkBlobOptions(blobOptions);
    } catch (e) {
        throw `(blobs2): ${e}`;
    }
    return renderPath(genFromOptions(blobOptions));
};
