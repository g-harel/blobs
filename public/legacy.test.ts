import blobs, {BlobOptions} from "./legacy";

const genMinimalOptions = (): BlobOptions => ({
    size: 1000 * Math.random(),
    complexity: 1 - Math.random(),
    contrast: 1 - Math.random(),
    color: "#fff",
});

describe("legacy", () => {
    it("should return a different result when seed is not provided", () => {
        const options = genMinimalOptions();

        const a = blobs(options);
        const b = blobs(options);

        expect(a).not.toEqual(b);
    });

    it("should return the same result when the seed is provided", () => {
        const options = genMinimalOptions();

        options.seed = "abcde";
        const a = blobs(options);
        const b = blobs(options);

        expect(a).toEqual(b);
    });

    it("should require options be provided", () => {
        expect(() => (blobs as any)()).toThrow("options");
    });

    it("should require a size be provided", () => {
        const options = genMinimalOptions();

        delete (options as any).size;
        expect(() => blobs(options)).toThrow("size");
    });

    it("should reject out of range complexity values", () => {
        const options = genMinimalOptions();

        options.complexity = 1234;
        expect(() => blobs(options)).toThrow("complexity");

        options.complexity = 0;
        expect(() => blobs(options)).toThrow("complexity");
    });

    it("should reject out of range contrast values", () => {
        const options = genMinimalOptions();

        options.contrast = 999;
        expect(() => blobs(options)).toThrow("contrast");

        options.contrast = -1;
        expect(() => blobs(options)).toThrow("contrast");
    });

    it("should reject options without stroke or color", () => {
        const options = genMinimalOptions();

        delete options.stroke;
        delete options.color;
        expect(() => blobs(options)).toThrow("stroke");
        expect(() => blobs(options)).toThrow("color");
    });
});

describe("editable", () => {
    it("should reflect changes when edited", () => {
        const options = genMinimalOptions();

        const out = blobs.editable(options);
        const initial = out.render();
        out.attributes.id = "test";
        const modified = out.render();

        expect(modified).not.toBe(initial);
    });
});

// Sanity checks to ensure the output remains consistent
// across changes to the source.
const testCases: Record<string, BlobOptions> = {
    fill: {
        size: 109,
        complexity: 0.1,
        contrast: 0.331,
        color: "red",
        seed: "fill",
    },
    stroke: {
        size: 226,
        complexity: 0.91,
        contrast: 0.6,
        stroke: {
            color: "#ff00bb",
            width: 3.8,
        },
        seed: "stroke",
    },
    guides: {
        size: 781,
        complexity: 1,
        contrast: 0.331,
        color: "yellow",
        guides: true,
        seed: "guides",
    },
};

for (const testCase of Object.keys(testCases)) {
    test(testCase, () => {
        expect(blobs(testCases[testCase])).toMatchSnapshot();
    });
}
