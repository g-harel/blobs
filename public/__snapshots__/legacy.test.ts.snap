// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fill 1`] = `"<svg width="109" height="109" viewBox="0 0 109 109" xmlns="http://www.w3.org/2000/svg"><g transform="rotate(29.498919679317623,54.5,54.5)"><path stroke="none" stroke-width="0" fill="red" d="M54.5,14.904644499132331C75.95401897738759,14.90464449913233,92.58177480526493,33.045981022612416,92.58177480526493,54.5C92.58177480526493,75.3898565168627,75.3898565168627,92.06532466605346,54.5,92.06532466605346C35.140079294728,92.06532466605346,22.151190205921836,73.859920705272,22.15119020592184,54.50000000000001C22.15119020592184,34.532642771227444,34.532642771227444,14.90464449913233,54.5,14.904644499132331"/></g></svg>"`;

exports[`guides 1`] = `"<svg width="781" height="781" viewBox="0 0 781 781" xmlns="http://www.w3.org/2000/svg"><g transform="rotate(3.442489226513049,390.5,390.5)"><path stroke="black" stroke-width="1" fill="yellow" d="M390.5,126.33058701145353C422.7599822291789,129.8894079821013,457.1517369455021,130.12745273744568,484.7560168794348,147.19717416039634C512.0337328801013,164.0649575005215,519.8607664658696,201.0400941880963,544.5059054044544,221.56356932561312C570.4261452389635,243.14889692825656,611.7577679542321,243.3499527852702,632.3240995385551,270.0859062179718C652.5414957686334,296.3682484004677,656.0291303230811,332.78264002578373,655.5644381263628,365.9381626741163C655.1113638087595,398.264756287615,638.8841165811418,427.5606398067945,629.6826289176689,458.55332532722855C620.5587046845698,489.2847597224773,615.7027870312934,520.9907034734381,601.0633639288316,549.5100713596294C585.2723467720305,580.2728838311288,570.3294797379046,615.2818420679225,540.3690947196874,632.5468987884551C510.32201395764275,649.8619150450116,472.07720239655697,638.2572446489619,437.8101304629824,643.5868442009801C404.43907720753396,648.777084873582,373.138427800321,665.5412734118242,339.41097189375944,663.8021610840965C304.54154473726527,662.0041644964895,267.47738691065854,655.071759972366,240.1072970361148,633.3925551389201C212.71820404244946,611.6982983815018,203.52330653817333,574.7971987713082,187.7077458632539,543.6416016528067C173.62525299120972,515.9000322349515,160.14941679690554,488.38496713511927,151.34885888571608,458.54436627473683C142.50320157704357,428.55084407312614,137.3614490164037,398.0976221087325,135.5767806585398,366.87788498406366C133.6378016292758,332.9587451522112,121.30298105955639,294.07541490260604,140.3383533379282,265.9343010505837C160.16298030584315,236.62638542805374,209.36566680771864,244.31364742916202,234.98836978172793,219.91186788626771C261.3342517552232,194.82136886934256,256.65497608331873,144.23139189508154,287.8278664113615,125.47267268305961C317.404723276248,107.67438874903971,356.1890316796786,122.54550763000411,390.5,126.33058701145353"/><line x1="390.5" y1="126.33058701145353" x2="422.7599822291789" y2="129.8894079821013" stroke-width="1" stroke="black"/><line x1="484.7560168794348" y1="147.19717416039634" x2="457.1517369455021" y2="130.12745273744568" stroke-width="1" stroke="black" stroke-dasharray="2"/><circle cx="422.7599822291789" cy="129.8894079821013" r="1" fill="black"/><circle cx="457.1517369455021" cy="130.12745273744568" r="1" fill="black"/><circle cx="390.5" cy="126.33058701145353" r="2" fill="black"/><line x1="484.7560168794348" y1="147.19717416039634" x2="512.0337328801013" y2="164.0649575005215" stroke-width="1" stroke="black"/><line x1="544.5059054044544" y1="221.56356932561312" x2="519.8607664658696" y2="201.0400941880963" stroke-width="1" stroke="black" stroke-dasharray="2"/><circle cx="512.0337328801013" cy="164.0649575005215" r="1" fill="black"/><circle cx="519.8607664658696" cy="201.0400941880963" r="1" fill="black"/><circle cx="484.7560168794348" cy="147.19717416039634" r="2" fill="black"/><line x1="544.5059054044544" y1="221.56356932561312" x2="570.4261452389635" y2="243.14889692825656" stroke-width="1" stroke="black"/><line x1="632.3240995385551" y1="270.0859062179718" x2="611.7577679542321" y2="243.3499527852702" stroke-width="1" stroke="black" stroke-dasharray="2"/><circle cx="570.4261452389635" cy="243.14889692825656" r="1" fill="black"/><circle cx="611.7577679542321" cy="243.3499527852702" r="1" fill="black"/><circle cx="544.5059054044544" cy="221.56356932561312" r="2" fill="black"/><line x1="632.3240995385551" y1="270.0859062179718" x2="652.5414957686334" y2="296.3682484004677" stroke-width="1" stroke="black"/><line x1="655.5644381263628" y1="365.9381626741163" x2="656.0291303230811" y2="332.78264002578373" stroke-width="1" stroke="black" stroke-dasharray="2"/><circle cx="652.5414957686334" cy="296.3682484004677" r="1" fill="black"/><circle cx="656.0291303230811" cy="332.78264002578373" r="1" fill="black"/><circle cx="632.3240995385551" cy="270.0859062179718" r="2" fill="black"/><line x1="655.5644381263628" y1="365.9381626741163" x2="655.1113638087595" y2="398.264756287615" stroke-width="1" stroke="black"/><line x1="629.6826289176689" y1="458.55332532722855" x2="638.8841165811418" y2="427.5606398067945" stroke-width="1" stroke="black" stroke-dasharray="2"/><circle cx="655.1113638087595" cy="398.264756287615" r="1" fill="black"/><circle cx="638.8841165811418" cy="427.5606398067945" r="1" fill="black"/><circle cx="655.5644381263628" cy="365.9381626741163" r="2" fill="black"/><line x1="629.6826289176689" y1="458.55332532722855" x2="620.5587046845698" y2="489.2847597224773" stroke-width="1" stroke="black"/><line x1="601.0633639288316" y1="549.5100713596294" x2="615.7027870312934" y2="520.9907034734381" stroke-width="1" stroke="black" stroke-dasharray="2"/><circle cx="620.5587046845698" cy="489.2847597224773" r="1" fill="black"/><circle cx="615.7027870312934" cy="520.9907034734381" r="1" fill="black"/><circle cx="629.6826289176689" cy="458.55332532722855" r="2" fill="black"/><line x1="601.0633639288316" y1="549.5100713596294" x2="585.2723467720305" y2="580.2728838311288" stroke-width="1" stroke="black"/><line x1="540.3690947196874" y1="632.5468987884551" x2="570.3294797379046" y2="615.2818420679225" stroke-width="1" stroke="black" stroke-dasharray="2"/><circle cx="585.2723467720305" cy="580.2728838311288" r="1" fill="black"/><circle cx="570.3294797379046" cy="615.2818420679225" r="1" fill="black"/><circle cx="601.0633639288316" cy="549.5100713596294" r="2" fill="black"/><line x1="540.3690947196874" y1="632.5468987884551" x2="510.32201395764275" y2="649.8619150450116" stroke-width="1" stroke="black"/><line x1="437.8101304629824" y1="643.5868442009801" x2="472.07720239655697" y2="638.2572446489619" stroke-width="1" stroke="black" stroke-dasharray="2"/><circle cx="510.32201395764275" cy="649.8619150450116" r="1" fill="black"/><circle cx="472.07720239655697" cy="638.2572446489619" r="1" fill="black"/><circle cx="540.3690947196874" cy="632.5468987884551" r="2" fill="black"/><line x1="437.8101304629824" y1="643.5868442009801" x2="404.43907720753396" y2="648.777084873582" stroke-width="1" stroke="black"/><line x1="339.41097189375944" y1="663.8021610840965" x2="373.138427800321" y2="665.5412734118242" stroke-width="1" stroke="black" stroke-dasharray="2"/><circle cx="404.43907720753396" cy="648.777084873582" r="1" fill="black"/><circle cx="373.138427800321" cy="665.5412734118242" r="1" fill="black"/><circle cx="437.8101304629824" cy="643.5868442009801" r="2" fill="black"/><line x1="339.41097189375944" y1="663.8021610840965" x2="304.54154473726527" y2="662.0041644964895" stroke-width="1" stroke="black"/><line x1="240.1072970361148" y1="633.3925551389201" x2="267.47738691065854" y2="655.071759972366" stroke-width="1" stroke="black" stroke-dasharray="2"/><circle cx="304.54154473726527" cy="662.0041644964895" r="1" fill="black"/><circle cx="267.47738691065854" cy="655.071759972366" r="1" fill="black"/><circle cx="339.41097189375944" cy="663.8021610840965" r="2" fill="black"/><line x1="240.1072970361148" y1="633.3925551389201" x2="212.71820404244946" y2="611.6982983815018" stroke-width="1" stroke="black"/><line x1="187.7077458632539" y1="543.6416016528067" x2="203.52330653817333" y2="574.7971987713082" stroke-width="1" stroke="black" stroke-dasharray="2"/><circle cx="212.71820404244946" cy="611.6982983815018" r="1" fill="black"/><circle cx="203.52330653817333" cy="574.7971987713082" r="1" fill="black"/><circle cx="240.1072970361148" cy="633.3925551389201" r="2" fill="black"/><line x1="187.7077458632539" y1="543.6416016528067" x2="173.62525299120972" y2="515.9000322349515" stroke-width="1" stroke="black"/><line x1="151.34885888571608" y1="458.54436627473683" x2="160.14941679690554" y2="488.38496713511927" stroke-width="1" stroke="black" stroke-dasharray="2"/><circle cx="173.62525299120972" cy="515.9000322349515" r="1" fill="black"/><circle cx="160.14941679690554" cy="488.38496713511927" r="1" fill="black"/><circle cx="187.7077458632539" cy="543.6416016528067" r="2" fill="black"/><line x1="151.34885888571608" y1="458.54436627473683" x2="142.50320157704357" y2="428.55084407312614" stroke-width="1" stroke="black"/><line x1="135.5767806585398" y1="366.87788498406366" x2="137.3614490164037" y2="398.0976221087325" stroke-width="1" stroke="black" stroke-dasharray="2"/><circle cx="142.50320157704357" cy="428.55084407312614" r="1" fill="black"/><circle cx="137.3614490164037" cy="398.0976221087325" r="1" fill="black"/><circle cx="151.34885888571608" cy="458.54436627473683" r="2" fill="black"/><line x1="135.5767806585398" y1="366.87788498406366" x2="133.6378016292758" y2="332.9587451522112" stroke-width="1" stroke="black"/><line x1="140.3383533379282" y1="265.9343010505837" x2="121.30298105955639" y2="294.07541490260604" stroke-width="1" stroke="black" stroke-dasharray="2"/><circle cx="133.6378016292758" cy="332.9587451522112" r="1" fill="black"/><circle cx="121.30298105955639" cy="294.07541490260604" r="1" fill="black"/><circle cx="135.5767806585398" cy="366.87788498406366" r="2" fill="black"/><line x1="140.3383533379282" y1="265.9343010505837" x2="160.16298030584315" y2="236.62638542805374" stroke-width="1" stroke="black"/><line x1="234.98836978172793" y1="219.91186788626771" x2="209.36566680771864" y2="244.31364742916202" stroke-width="1" stroke="black" stroke-dasharray="2"/><circle cx="160.16298030584315" cy="236.62638542805374" r="1" fill="black"/><circle cx="209.36566680771864" cy="244.31364742916202" r="1" fill="black"/><circle cx="140.3383533379282" cy="265.9343010505837" r="2" fill="black"/><line x1="234.98836978172793" y1="219.91186788626771" x2="261.3342517552232" y2="194.82136886934256" stroke-width="1" stroke="black"/><line x1="287.8278664113615" y1="125.47267268305961" x2="256.65497608331873" y2="144.23139189508154" stroke-width="1" stroke="black" stroke-dasharray="2"/><circle cx="261.3342517552232" cy="194.82136886934256" r="1" fill="black"/><circle cx="256.65497608331873" cy="144.23139189508154" r="1" fill="black"/><circle cx="234.98836978172793" cy="219.91186788626771" r="2" fill="black"/><line x1="287.8278664113615" y1="125.47267268305961" x2="317.404723276248" y2="107.67438874903971" stroke-width="1" stroke="black"/><line x1="390.5" y1="126.33058701145353" x2="356.1890316796786" y2="122.54550763000411" stroke-width="1" stroke="black" stroke-dasharray="2"/><circle cx="317.404723276248" cy="107.67438874903971" r="1" fill="black"/><circle cx="356.1890316796786" cy="122.54550763000411" r="1" fill="black"/><circle cx="287.8278664113615" cy="125.47267268305961" r="2" fill="black"/></g></svg>"`;

exports[`stroke 1`] = `"<svg width="226" height="226" viewBox="0 0 226 226" xmlns="http://www.w3.org/2000/svg"><g transform="rotate(10.625043198466301,113,113)"><path stroke="#ff00bb" stroke-width="3.8" fill="none" d="M113,42.187082298945086C122.79474494878556,40.93994738588087,128.08514500519405,54.842942154431036,137.1295411257755,58.80416329407413C146.24165581029774,62.79504348485257,157.51842254183055,60.01704955511825,165.90945623567904,65.36011162379504C174.71687671447282,70.96831151907477,181.0880692974371,79.95796015403897,185.33243631773502,89.49776676385548C189.73428003886485,99.39152421352725,196.20569839281566,111.822474823727,190.74027198902974,121.17083183675264C184.275146635445,132.22912728784644,166.3394780066649,129.73602552824286,156.80578145715737,138.2912797163517C150.81862065087725,143.66397878400295,152.63580758814572,154.1776323719105,147.27490184117337,160.17535522011516C141.64626912802416,166.4726080764768,133.5690193595429,169.81059568769254,125.75894924668432,173.0261367912584C117.04502915031891,176.6138085839763,108.09560947846845,179.1793135116963,98.72369653099743,180.1647271523213C87.32619013379446,181.36312264914937,72.49281709265853,187.87055519411282,64.72599173408246,179.44347220588492C56.10054832930125,170.08478007279948,66.09720211216099,154.1178572037905,63.27873223580451,141.70658732810776C61.3893271033385,133.38649946758036,50.99025921153393,128.03545137040612,51.11446122271466,119.50443222719099C51.241275506902305,110.79398482176407,63.010835822466824,105.71867007151292,63.928797313315314,97.05579972925904C65.17258032291737,85.31812766621682,49.49602866888215,71.69698855511481,57.133958677847595,62.69799045462378C64.63143952592921,53.86446935255059,80.5830156751293,68.61242545584774,91.45947685068577,64.6191928775053C101.29841121040204,61.00688253251097,102.60284327876242,43.51092046027091,113,42.187082298945086"/></g></svg>"`;
